src/main/java/com/redhat/devtools/lsp4ij/internal/CompletableFutures.java

```diff
@@ -109,6 +110,23 @@ public static void waitUntilDone(@Nullable CompletableFuture<?> future,
         if (future == null) {
             return;
         }
+        waitUntilDoneOrTimeout(future, file, 25, null);
+    }
+    /**
+     * Wait for the given future to complete and stop the wait if {@link ProcessCanceledException} is thrown.
+     * If the timeout is reached, the future is put into the <code>pendingFutures</code> list and the method returns false.
+     * If <code>pendingFutures</code> is null, the function will retry the wait indefinitely, blocking the thread.
+     *
+     * @param future         the future to wait.
+     * @param file           the Psi file.
+     * @param timeout        the amount of milliseconds to wait in {@link Future#get(long, TimeUnit)}
+     * @param pendingFutures the list to put the unfinished future into on timeout. If <code>null</code>,
+     *                       the timeout is retried indefinitely until successful.
+     */
+    public static boolean waitUntilDoneOrTimeout(@NotNull CompletableFuture<?> future,

```

Comment:
Instead of having a new method, I think we should add an Integer timeout which can be null.

pendingFutures is a specific case that LSP inlay hints should manage.